File: Java/A-PREP/Code Signal Industry/In-Memory Database/level3.java
import java.util.*;

public class InMemoryDatabase {
    // Map to store records (key -> {field -> {value, expirationTimestamp}})
    private Map<String, Map<String, FieldValue>> database;

    // Constructor to initialize the database
    public InMemoryDatabase() {
        database = new HashMap<>();
    }

    // Inner class to store value and expiration timestamp
    private static class FieldValue {
        String value;
        long expirationTimestamp; // -1 means no expiration

        FieldValue(String value, long expirationTimestamp) {
            this.value = value;
            this.expirationTimestamp = expirationTimestamp;
        }
    }

    // Method to insert or update a field-value pair with a timestamp
    public String setAt(String key, String field, String value, long timestamp) {
        database.putIfAbsent(key, new HashMap<>());
        database.get(key).put(field, new FieldValue(value, -1)); // No expiration
        return "";
    }

    // Method to insert or update a field-value pair with a timestamp and TTL
    public String setAtWithTTL(String key, String field, String value, long timestamp, long ttl) {
        database.putIfAbsent(key, new HashMap<>());
        database.get(key).put(field, new FieldValue(value, timestamp + ttl));
        return "";
    }

    // Method to delete a field at a specific timestamp
    public String deleteAt(String key, String field, long timestamp) {
        if (!database.containsKey(key)) {
            return "false";
        }
        Map<String, FieldValue> record = database.get(key);
        if (!record.containsKey(field)) {
            return "false";
        }
        record.remove(field);
        return "true";
    }

    // Method to get the value of a field at a specific timestamp
    public String getAt(String key, String field, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        if (!record.containsKey(field)) {
            return "";
        }
        FieldValue fieldValue = record.get(field);
        if (fieldValue.expirationTimestamp != -1 && timestamp >= fieldValue.expirationTimestamp) {
            return "";
        }
        return fieldValue.value;
    }

    // Method to scan all fields of a record at a specific timestamp
    public String scanAt(String key, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        // Filter out expired fields
        Map<String, String> validFields = new HashMap<>();
        for (Map.Entry<String, FieldValue> entry : record.entrySet()) {
            if (entry.getValue().expirationTimestamp == -1 || timestamp < entry.getValue().expirationTimestamp) {
                validFields.put(entry.getKey(), entry.getValue().value);
            }
        }
        // Sort the valid fields lexicographically
        TreeMap<String, String> sortedFields = new TreeMap<>(validFields);
        // Build the result string
        return buildResultString(sortedFields);
    }

    // Method to scan fields of a record that start with a prefix at a specific timestamp
    public String scanByPrefixAt(String key, String prefix, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        // Filter out expired fields and fields that don't start with the prefix
        Map<String, String> validFields = new HashMap<>();
        for (Map.Entry<String, FieldValue> entry : record.entrySet()) {
            if ((entry.getValue().expirationTimestamp == -1 || timestamp < entry.getValue().expirationTimestamp) &&
                entry.getKey().startsWith(prefix)) {
                validFields.put(entry.getKey(), entry.getValue().value);
            }
        }
        // Sort the valid fields lexicographically
        TreeMap<String, String> sortedFields = new TreeMap<>(validFields);
        // Build the result string
        return buildResultString(sortedFields);
    }

    // Helper method to build the result string
    private String buildResultString(Map<String, String> map) {
        StringBuilder result = new StringBuilder();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            if (result.length() > 0) {
                result.append(", ");
            }
            result.append(entry.getKey()).append("(").append(entry.getValue()).append(")");
        }
        return result.toString();
    }

    // Main method to test the functionality
    public static void main(String[] args) {
        InMemoryDatabase db = new InMemoryDatabase();

        // Test case 1: SET_AT_WITH_TTL operation
        System.out.println(db.setAtWithTTL("A", "BC", "E", 1, 9)); // ""
        System.out.println(db.setAtWithTTL("A", "BC", "E", 5, 10)); // ""

        // Test case 2: SET_AT operation
        System.out.println(db.setAt("A", "BD", "F", 5)); // ""

        // Test case 3: SCAN_BY_PREFIX_AT operation
        System.out.println(db.scanByPrefixAt("A", "B", 14)); // "BC(E), BD(F)"
        System.out.println(db.scanByPrefixAt("A", "B", 15)); // "BD(F)"
    }
}

File: Python/CodeSignal/Industry Framework/In-Memory Storage/level3.py
class InMemoryDatabase:
    def __init__(self):
        self.records = {}  # key -> {field: (value, expiry_time or None)}
    
    def _cleanup_expired(self, timestamp: int):
        """Remove expired fields"""
        keys_to_remove = []
        for key in self.records:
            fields_to_remove = []
            for field, (value, expiry) in self.records[key].items():
                if expiry is not None and timestamp >= expiry:
                    fields_to_remove.append(field)
            
            for field in fields_to_remove:
                del self.records[key][field]
            
            if not self.records[key]:
                keys_to_remove.append(key)
        
        for key in keys_to_remove:
            del self.records[key]
    
    # Level 1 operations (backward compatibility)
    def set(self, key: str, field: str, value: str) -> str:
        if key not in self.records:
            self.records[key] = {}
        
        self.records[key][field] = (value, None)  # No expiry
        return ""
    
    def get(self, key: str, field: str) -> str:
        if key not in self.records or field not in self.records[key]:
            return ""
        
        value, expiry = self.records[key][field]
        return value
    
    def delete(self, key: str, field: str) -> str:
        if key not in self.records or field not in self.records[key]:
            return "false"
        
        del self.records[key][field]
        
        if not self.records[key]:
            del self.records[key]
        
        return "true"
    
    def scan(self, key: str) -> str:
        if key not in self.records:
            return ""
        
        fields = sorted(self.records[key].keys())
        result_parts = []
        for field in fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    def scan_by_prefix(self, key: str, prefix: str) -> str:
        if key not in self.records:
            return ""
        
        matching_fields = []
        for field in self.records[key]:
            if field.startswith(prefix):
                matching_fields.append(field)
        
        matching_fields.sort()
        result_parts = []
        for field in matching_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    # Level 3 operations with timestamp
    def set_at(self, key: str, field: str, value: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            self.records[key] = {}
        
        self.records[key][field] = (value, None)  # No expiry
        return ""
    
    def set_at_with_ttl(self, key: str, field: str, value: str, timestamp: int, ttl: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            self.records[key] = {}
        
        expiry_time = timestamp + ttl
        self.records[key][field] = (value, expiry_time)
        return ""
    
    def get_at(self, key: str, field: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records or field not in self.records[key]:
            return ""
        
        value, expiry = self.records[key][field]
        if expiry is not None and timestamp >= expiry:
            return ""
        
        return value
    
    def delete_at(self, key: str, field: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records or field not in self.records[key]:
            return "false"
        
        value, expiry = self.records[key][field]
        if expiry is not None and timestamp >= expiry:
            return "false"
        
        del self.records[key][field]
        
        if not self.records[key]:
            del self.records[key]
        
        return "true"
    
    def scan_at(self, key: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            return ""
        
        valid_fields = []
        for field, (value, expiry) in self.records[key].items():
            if expiry is None or timestamp < expiry:
                valid_fields.append(field)
        
        valid_fields.sort()
        result_parts = []
        for field in valid_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    def scan_by_prefix_at(self, key: str, prefix: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            return ""
        
        matching_fields = []
        for field, (value, expiry) in self.records[key].items():
            if field.startswith(prefix) and (expiry is None or timestamp < expiry):
                matching_fields.append(field)
        
        matching_fields.sort()
        result_parts = []
        for field in matching_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)

def main():
    db = InMemoryDatabase()
    
    # Test cases from level 3 example
    queries = [
        ["SET_AT_WITH_TTL", "A", "BC", "E", "1", "9"],
        ["SET_AT_WITH_TTL", "A", "BC", "E", "5", "10"],
        ["SET_AT", "A", "BD", "F", "5"],
        ["SCAN_BY_PREFIX_AT", "A", "B", "14"],
        ["SCAN_BY_PREFIX_AT", "A", "B", "15"]
    ]
    
    results = []
    for query in queries:
        operation = query[0]
        if operation == "SET":
            result = db.set(query[1], query[2], query[3])
        elif operation == "GET":
            result = db.get(query[1], query[2])
        elif operation == "DELETE":
            result = db.delete(query[1], query[2])
        elif operation == "SCAN":
            result = db.scan(query[1])
        elif operation == "SCAN_BY_PREFIX":
            result = db.scan_by_prefix(query[1], query[2])
        elif operation == "SET_AT":
            result = db.set_at(query[1], query[2], query[3], int(query[4]))
        elif operation == "SET_AT_WITH_TTL":
            result = db.set_at_with_ttl(query[1], query[2], query[3], int(query[4]), int(query[5]))
        elif operation == "GET_AT":
            result = db.get_at(query[1], query[2], int(query[3]))
        elif operation == "DELETE_AT":
            result = db.delete_at(query[1], query[2], int(query[3]))
        elif operation == "SCAN_AT":
            result = db.scan_at(query[1], int(query[2]))
        elif operation == "SCAN_BY_PREFIX_AT":
            result = db.scan_by_prefix_at(query[1], query[2], int(query[3]))
        
        results.append(result)
        print(f"{query}: {result}")
    
    print(f"Final results: {results}")
    # Expected: ["", "", "", "BC(E), BD(F)", "BD(F)"]

if __name__ == "__main__":
    main()